dspy/utils/usage_tracker.py

```diff
@@ -0,0 +1,59 @@
+"""Usage tracking utilities for DSPy."""
+
+from collections import defaultdict
+from contextlib import contextmanager
+from typing import Any
+
+from dspy.dsp.utils.settings import settings
+
+
+class UsageTracker:
+    """Tracks LM usage data within a context."""
+
+    def __init__(self):
+        self.usage_data = defaultdict(list)
+
+    def _flatten_usage_entry(self, usage_entry) -> dict[str, dict[str, Any]]:
+        result = dict(usage_entry)
+        result["completion_tokens_details"] = dict(result["completion_tokens_details"])
+        result["prompt_tokens_details"] = dict(result["prompt_tokens_details"])
+        return result
+
+    def _merge_usage_entries(self, usage_entry1, usage_entry2) -> dict[str, dict[str, Any]]:
+        if usage_entry1 is None or len(usage_entry1) == 0:
+            return dict(usage_entry2)
+        if usage_entry2 is None or len(usage_entry2) == 0:
+            return dict(usage_entry1)
+
+        result = dict(usage_entry2)
+        for k, v in usage_entry1.items():
+            if k in result:
+                if isinstance(v, dict):
+                    result[k] = self._merge_usage_entries(result[k], v)
+                else:
+                    result[k] = result[k] or 0
+                    result[k] += v if v else 0
+        return result
+
+    def add_usage(self, lm: str, usage_entry: dict):
```

Comment:
nit: shall we use `model` instead of lm? In DSPy codebase, I see lm is mostly used for dspy.LM object

Reply:
got your concern. It's actually meaning the string representation of `dspy.settings.LM`, so I kinda prefer the name `lm` here for consistency.